File: leetcode/Search_Insert_Position.java
import java.util.*;

public class Solution {
    public int searchInsert(int[] A, int target) {
        if (target <= A[0]) return 0;
        if (target > A[A.length - 1]) return A.length;
        int a = 0;
        int b = A.length - 1;
        while (a + 1 < b) {
            int m = (a + b) / 2;
            if (A[m] >= target) b = m;
            else a = m;
        }
        return b;
    }
}

File: leetcode/Search_Insert_Position.cpp
class Solution {
    int binarySearch(int A[], int beg, int end, int target) {
        if (beg + 1 == end) {
            if (A[beg] >= target)
                return beg - 1;
            else
                return beg;
        }
        int mid = (beg + end) / 2;
        if (A[mid] >= target)
            return binarySearch(A, beg, mid, target);
        else
            return binarySearch(A, mid, end, target);
    }
public:
    int searchInsert(int A[], int n, int target) {
        return binarySearch(A, 0, n, target) + 1;
    }
};
